File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int h = triangle.length;
        
        int[][] dp = new int[h][h];
        dp[0][0] = triangle[0][0];
        
        for (int i = 1; i < h; i++) {
            for (int j = 0; j < i+1; j++) {
                if (j == 0) {
                    dp[i][j] = dp[i-1][0] + triangle[i][j];
                } else if (j == i) {
                    dp[i][j] = dp[i-1][j-1] + triangle[i][j];
                } else {
                    dp[i][j] = Math.max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j];
                }
            } 
        }
        
        // 삼각형의 바닥에서 가장 큰 값 구하기
        for(int i = 0; i < h; i++) {
            answer = Math.max(dp[h-1][i], answer);
        }
        return answer;
    }
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    answer = 0
    
    dp = [[] for _ in range(len(triangle))] # 해당 칸의 최대값 저장 dp 테이블
    
    dp[0].append(triangle[0][0])
    
    for i in range(1, len(triangle)):
        for j in range(len(triangle[i])):
            if j == 0:
                dp[i].append(dp[i-1][0] + triangle[i][j])
            elif j == len(triangle[i]) - 1:
                dp[i].append(dp[i-1][-1] + triangle[i][j])
            else:
                dp[i].append(max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j])
    
    answer = max(dp[-1])
    return answer